oak_functions/loader/src/tf.rs

```diff
@@ -0,0 +1,104 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::Context;
+use bytes::{Buf, Bytes};
+use hyper::{Body, Client, Request};
+use hyper_rustls::HttpsConnector;
+use oak_functions_abi::proto::Inference;
+use tract_tensorflow::prelude::*;
+type Model = RunnableModel<TypedFact, Box<dyn TypedOp>, Graph<TypedFact, Box<dyn TypedOp>>>;
+
+pub struct TensorFlowModel {
+    model: Model,
+    pub shape: Vec<u8>,
+}
+
+impl TensorFlowModel {
+    /// Creates an instance of TensorFlowModel, by loading the model in the specified path, and
+    /// configuring it with the given shape.
+    pub fn create(bytes: Bytes, shape: Vec<u8>) -> anyhow::Result<Self> {
+        let dim = shape
+            .iter()
+            .cloned()
+            .map(|u| u.into())
+            .collect::<Vec<i32>>();
+        let mut reader = bytes.reader();
+
+        // Creates the model based on the example in https://github.com/sonos/tract/tree/main/examples/tensorflow-mobilenet-v2
+        let model = tract_tensorflow::tensorflow()
+            // load the model
+            .model_for_read(&mut reader)?
+            // specify input type and shape
+            .with_input_fact(0, InferenceFact::dt_shape(f32::datum_type(), dim))?
+            // optimize the model
+            .into_optimized()?
+            // make the model runnable and fix its inputs and outputs
+            .into_runnable()?;
+
+        Ok(TensorFlowModel { model, shape })
+    }
+
+    /// Run the model for the given input and return the resulting inference vector.
+    pub fn get_inference(&self, tensor_bytes: &[u8], tensor_shape: &[usize]) -> Inference {
+        // Reshape the input
+        let tensor: Tensor =
+            unsafe { Tensor::from_raw::<f32>(tensor_shape, tensor_bytes).unwrap() };
```

Comment:
What are the potential consequenses of a malformed Tensor? Are there any potential security risks if the Wasm code sends bad data? Is there some validation that can be done before running unsafe code that parses untrusted data?

Reply:
I had a look at their code, and I think it is ok. The raw bytes are only used to update the `data` field on the Tensor that is created from the shape. I was worried that the code might doing more complex unsafe casting.